examples/perf/ipc.py

```diff
@@ -0,0 +1,137 @@
+#!/usr/bin/python
+# IPC - Instructions Per Cycles using Perf Events and
+# uprobes
+# 24-Apr-2020	Saleem Ahmad	Created this.
+
+from __future__ import print_function
+from bcc import BPF, USDT, Perf
+from bcc.utils import printb
+from time import sleep
+import sys
+
+# load BPF program
+code="""
+#include <uapi/linux/ptrace.h>
+
+struct perf_delta {
+    u64 clk_delta;
+    u64 inst_delta;
+    u64 time_delta;
+};
+
+// Perf Arrays to read counter values for open
+// perf events.
+BPF_PERF_ARRAY(clk, 16);
+BPF_PERF_ARRAY(inst, 16);
+
+// Perf Output
+BPF_PERF_OUTPUT(output);
+
+// Per Cpu Data to store start values
+BPF_PERCPU_ARRAY(data, u64);
+
+void trace_start(struct pt_regs *ctx) {
+    u32 clk_k = 0;
+    u32 inst_k = 1;
+    u32 time = 2;
+
+    int cpu = bpf_get_smp_processor_id();
+    u64 clk_start = clk.perf_read(cpu);
+    u64 inst_start = inst.perf_read(cpu);
+    u64 time_start = bpf_ktime_get_ns();
+    
+    u64* kptr = NULL;
+    kptr = data.lookup(&clk_k);
+    if (kptr) {
+        data.update(&clk_k, &clk_start);
+    } else {
+        data.insert(&clk_k, &clk_start);
+    }
+
+    kptr = data.lookup(&inst_k);
+    if (kptr) {
+        data.update(&inst_k, &inst_start);
+    } else {
+        data.insert(&inst_k, &inst_start);
+    }
+
+    kptr = data.lookup(&time);
+    if (kptr) {
+        data.update(&time, &time_start);
+    } else {
+        data.insert(&time, &time_start);
+    }
+}
+
+void trace_end(struct pt_regs* ctx) {
+    u32 clk_k = 0;
+    u32 inst_k = 1;
+    u32 time = 2;
+
+    int cpu = bpf_get_smp_processor_id();
+    u64 clk_end = clk.perf_read(cpu);
+    u64 inst_end = inst.perf_read(cpu);
+    u64 time_end = bpf_ktime_get_ns();
+    
+    struct perf_delta perf_data = {} ;
+    bool submit = true;
+    u64* kptr = NULL;
+    kptr = data.lookup(&clk_k);
+    if (kptr) {
+        perf_data.clk_delta = clk_end - *kptr;
+    }
+    
+    kptr = data.lookup(&inst_k);
+    if (kptr) {
+        perf_data.inst_delta = inst_end - *kptr;
+    } else {
+        submit = false;
+    }
+
+    kptr = data.lookup(&time);
+    if (kptr) {
+        perf_data.time_delta = time_end - *kptr;
+    } else {
+        submit = false;
+    }
+
+    if (submit) {
+        output.perf_submit(ctx, &perf_data, sizeof(struct perf_delta));
+    }
+}
+"""
+
+b = BPF(text=code)
+
+# Attach Probes at start and end of the trace function
+b.attach_uprobe(name="c", sym="strlen", fn_name="trace_start")
+b.attach_uretprobe(name="c", sym="strlen", fn_name="trace_end")
```

Comment:
In my system, I have
```
-bash-4.4$ nm /usr/lib64/libc.so.6 | grep strlen
000000000008cd80 t __GI_strlen
000000000008cd30 i strlen
000000000008cd80 t __strlen_sse2
000000000016c3d0 t __strlen_sse2_no_bsf
000000000016f630 t __strlen_sse2_pminub
0000000000146cc0 t __strlen_sse42
```
Read the `nm` man page, we have
```
           "i" For PE format files this indicates that the symbol is in a section specific to the implementation of
               DLLs.  For ELF format files this indicates that the symbol is an indirect function.  This is a GNU
               extension to the standard set of ELF symbol types.  It indicates a symbol which if referenced by a
               relocation does not evaluate to its address, but instead must be invoked at runtime.  The runtime
               execution will then return the value to be used in the relocation.
```

That is, during runtime relocation stage  by runtime linker, `strlen` is called once to returns a different address
for example `__strlen_sse2` and later calling strlen from application will call `__strlen_sse2` instead. 

What is your local libc `strlen` definition looks like? We should add some comments here  to warn people `strlen` may not work depending on libc.

So this overhead may be big enough compared to actual strlen. Only big input for strlen may have a meaningful output, I think. Please add some comments along this line to warn potential users about how to interpret the results.

Reply:
My strlen definition looks like following

```bash
➜  ~ nm -D /lib/x86_64-linux-gnu/libc-2.27.so | egrep strlen
000000000009dc70 i strlen
```

> Please add some comments along this line to warn potential users about how to interpret the results.
Sure, I will add extra comments about this and also about the context switch from user space to kernel space.

I think since its generic IPC script, better to take library name and function to trace as commnad line argument instead of just calculating for `strlen`.

Reply:
@yonghong-song I have updated the script with notes on result. I have also added command line argument to run the script on any function inside a library.